pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/RecordInfo.java

```diff
@@ -18,19 +18,27 @@
  */
 package org.apache.pinot.segment.local.upsert;
 
+import javax.annotation.Nullable;
 import org.apache.pinot.spi.data.readers.PrimaryKey;
 
 
 @SuppressWarnings("rawtypes")
 public class RecordInfo {
+  // This default exists for backward compatibility, it is not a real virtual column but uses the $-prefix so that
+  // it cannot conflict with a real column
+//  static final String DEFAULT_COMP_COL_KEY = "$comparisonColumn";
   private final PrimaryKey _primaryKey;
   private final int _docId;
   private final Comparable _comparisonValue;
+  private final String _comparisonColumnName;
```

Comment:
Adding this field will add extra memory footprint to the existing use cases. I think this info can be derived from the `_comparisonValue`, and we should add a special comparator for the new `_comparisonValue`